### DISCUSSION TITLE
No semantic tokens for many keywords

### DISCUSSION THREAD

**Discussion by HoldMyBeer1111** [mrcjkb/rustaceanvim#503] [Q&A]
Hi, why do some keywords (eg `for`) have `Semantic Tokens`, but others dont? Any ideas?

`:Inspect` for `for`
![image](https://github.com/user-attachments/assets/e92268d2-943c-4908-87d4-f11fc90e5a61)

`:Inspect` for `let`
![image](https://github.com/user-attachments/assets/21699e9d-bd6c-4ee0-97c8-b9f2ece5f618)

**Comment by mrcjkb** [comment]
Hey :wave:

that would be a [rust-analyzer question](https://github.com/rust-lang/rust-analyzer/discussions).

I think many LSP servers don't provide semantic tokens for *all* keywords.

**Reply by HoldMyBeer1111** [reply]
But it works properly in vscode
![image](https://github.com/user-attachments/assets/17b0acb9-1cbf-4328-a6ba-5d32a72db09f)

**Reply by mrcjkb** [reply]
then it's likely either

- Your colorscheme not supporting all tokens.
- Use of different rust-analyzer builds in Neovim and VSCode.

**Reply by HoldMyBeer1111** [reply]
> * Your colorscheme not supporting all tokens.

wdym? I have `vim.api.nvim_set_hl(0, '@lsp.type.keyword.rust', { fg = '#d3869b' })` in a config file, and it doesnt work because some keywords lack semantic tokens from LSP. Looks like nothing to do with a colorscheme.

> * Use of different rust-analyzer builds in Neovim and VSCode.

I have checked with the same binary.
vscode
![image](https://github.com/user-attachments/assets/4b8117af-70ae-4273-9369-b50185a56e45)
neovim
![image](https://github.com/user-attachments/assets/2a321944-c178-45fc-84b2-a6290937f6e8)

**Reply by mrcjkb** [reply]
Can you confirm that the highlighting of the `let` keyword in VSCode in fact comes from rust-analyzer and is not part of VSCode's theming?
[The documentation](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#theming), the texmate scopes listed in your screenshot, and [this file](https://github.com/microsoft/vscode/blob/bfc1636daf4c2b315107316e38a0498f7f0ef81d/extensions/rust/syntaxes/rust.tmLanguage.json#L1), lead me to believe that might not be the case.
In VSCode, "semantic token" is not necessarily synonymous with LSP semantic tokens.

rustaceanvim doesn't implement any highlighting or theming, which is out of scope for this plugin.

**Reply by HoldMyBeer1111** [reply]
`let` is colored differently and does not have a "semantic token type" if rust-analyzer extension is disabled.

**Reply by mrcjkb** [reply]
Then I suggest asking in the rust-analyzer discussions. It could be a non-standard token that isn't supported by Neovim or a server configuration issue.

As I said, highlighting is out of scope for rustaceanvim.
